### ISSUE TITLE
axis_geo() development

### ISSUE THREAD

**Issue by KEichenseer** [palaeoverse-community/palaeoverse#1]
Functionality of the function

- [x] Plotting a period bar
- [x] Plotting a stage bar
- [x] Plotting a joint period + stage bar
- [x] manual age limits input
- [x] option for stage / period colours (default to white bar with black labels)
- [ ] intelligent label lengths
- [ ] intelligent switching between periods and stages if not specified
- [ ] intelligent bar widths (adjusting to the plotting window size?)
- [x] documentation
- [x] tests

**Comment by willgearty** [comment]
I've made a new function to accomplish this in the `axis-geo` branch. It can plot as many stacked scales as you want on a single side. Basically, I tried to make it work as much like `deeptime::coord_geo` as possible (so almost all of the arguments are the same). It doesn't do any of the "intelligent" stuff mentioned in the list above, but let me know if you think any of those are still required given the new functionality I've implemented @KEichenseer. The documentation is all done, but I haven't written any tests yet, but those should be fairly straightforward.

**Comment by willgearty** [comment]
Tests are done!

**Comment by LewisAJones** [comment]
Small request for updating the name of the function to time_plot to follow the naming convention of other functions time_bins.R, time_binning.R, lat_bins.R, lat_plot.R, etc.

What do you think?

**Comment by willgearty** [comment]
I strongly disagree. With the new function, you aren't plotting any data. Rather, you are adding an axis, so axis_geo() is much more descriptive (and aligns with the base R axis() function).

**Comment by LewisAJones** [comment]
Sorry Will, misunderstanding here. I do not mean to change the name of your function axis_geo(), but the original function that Kilian was writing (tplot.R). Perhaps I have missed something, are you no longer having this function to generate a generate time-series plot and only using the axis_geo() to add to a base R plot after plotting (similar to your deeptime/ggplot implementation?). If the latter, happy with axis_geo().

**Comment by willgearty** [comment]
My plan was to remove the plotting function, since there are already (better) functions to plot data in base R.

**Comment by KEichenseer** [comment]
Hi Will, great job on the function, it works really well already! 

I did some testing (not on all the functionalities yet), and these are the issues I found (some of them may be fine as is, but we can discuss them here):

- [ ] old axis remains. Should the axis of the previous plot be eradicated by default, e.g., by covering it with a white polygon? This saves the user from having to turn off the corresponding axis every time.
- [ ] Axis labels overlap. Do we want to implement some automated label shortening / omission?
- [ ] trailing zeroes on axis ages should be removed.
- [ ] Do we want to round axis ages?
- [ ] no axis shown when ages are negative. Should we add a warning to say that the ages need to be positive / indicate the range of ages that is covered?
- [ ] rename "dat" argument to "interval" or "unit"?
- [ ] plotting multiple bars requires list input to the "dat" argument. Allow also vector input?
- [ ] should we allow for plotting axes along multiple sides at once? (vector or list input)
- [ ] Bar label position should be adjusted to show in the middle of the actually plotted part, otherwise labels are cut off/ pushed out
- [ ] Make white fill colour the default? Personally I find the colours distracting, but of course we retain the ability to switch on colours

Once we have settled on what to include, I'm happy to work on implementing these.

**Comment by willgearty** [comment]
Hey Killian, thanks for the ideas!

I grappled a bit with trying to make `axis_geo()` work just like `coord_geo()`, but I also wanted to make it feel like an extension of the built-in `axis()` function. To maintain the latter intent, I think we should not eradicate existing axes (# 1) and should only allow plotting an axis on a single side at a time (# 2) (despite the ability to plot multiple axes in `coord_geo()`). The normal `axis()` function works in this way, so I think we should emulate that.

For # 3 and # 4, you should be able to achieve a scale with negative axis values by specifying `neg = TRUE`, while you should be able to center the labels for intervals at the end of the scale using `center_end_labels = TRUE`.

For # 5, because some of the arguments can be vectors (e.g., `skip`), we need to use lists (vectors of vectors don't work in R), I'd prefer to have all of the arguments work the same way, so I'd prefer to have them all either be individual values (or vectors) for single scales and lists for multiple scales.

For # 6, it's tricky, because it can also be a data.frame with custom data, so I feel like `dat` is generic enough to cover both the string and data.frame cases?

For # 7 and # 8: It looks like the `axis()` function adds trailing zeros by default. I think (but may be incorrect) that the only way to avoid this is to specify `labels` in `axis()`. As it is currently designed, the user should be able to specify the `at` and `labels` arguments and they will be supplied to `axis()`, which I think is a nice feature, so we could just put leave it up to the user? But I think it would also be fine to at least always drop the trailing zeros (but ideally we would maintain the ability to specify `at` and `labels` for `axis`). We could also add a `round` argument, too, to round the labels for the axis ticks.

For # 9, do you mean instead of using the `colors` column in the data? Or instead of the grayscale that is used by default if no colors are supplied? In my experience, people really like the official GTS colors, and I see them used in most if not all publications that include timescales on plots. However, I'm fine with using white instead of the greyscale if no colors are provided.

Finally, for # 10, I think this would be amazing, but I would have no idea where to start. Feel free to give it a shot!

**Comment by KEichenseer** [comment]
Hi Will,

thanks! 

For #1 and #2, Ok, let's keep it as it is.

#3: I missed that, should we implement that this is automatically turned on if all values are negative?

# 4: That's awesome, I suggest to set the default to TRUE.

# 5: Ok, makes sense.

# 6: I see... the custom data would still be a data.frame of intervals, though? If it's always some sort of interval / time data, I'd still prefer a name that reflects that.

# 7 and 8: Speaking from experience, it is quite frustrating to manually change axis labels in multiple plots. Specifying 'at' and 'labels' we should keep, yes. Maybe we can include an automatic option to find even intervals and label them (e.g. 50, 100, 150 Ma) instead of the uneven interval boundaries. That's what I used in my publications, and I had implemented that in the earlier 'tplot()' function.

# 9: If the user specifies column data, that should be used. But if he specifies data without colours or just e.g. "periods", I'd opt for black & white. But if most prefer to use colours, we can keep it as is.

# 10: Ok I'll have a go at that.

**Comment by willgearty** [comment]
- 3: Sure, but only if all values are negative. if the axis has negative and positive values, it is hard to discern the intent of the user.
- 4: Sure, go ahead! But yes, if the interval is very narrow, this behavior probably isn't desirable.
- 6: I'd be fine with renaming it to `intervals` if you'd rather that. Make sure to update the documentation accordingly!
- 7/8: How about by default we don't specify at/labels at all and let axis() do it's thing. Then, we can add an argument for users to request that exact interval ages are used (`exact` or something like that), in which case we would use the current default behavior. Finally, users can always specify `at` and `labels` for even more customization.

**Comment by KEichenseer** [comment]
I implemented points 3 - 8. Point 10 is indeed tricky. One thought I had is to omit the label in cases where the bar segment width is below a certain percentage of the total plot height or width.

**Comment by willgearty** [comment]
Great, thanks! I'll take a look at the code and give it a thought tomorrow!

**Comment by willgearty** [comment]
OK, all of your changes look good to me @KEichenseer! I changed `interval` to `intervals` and did some minor tidying of the code, documentation, and tests.

I think whatever choice we make about point 10 won't satisfy all users (and could confuse many), so I suggest we think about it more and implement something in the future (or wait until someone else solves automated label sizing like in ggfittext but for base R).

Given that everything else is finished, I've submitted a pull request for review (#13)!
